src/sessions/PageManager.ts

```diff
@@ -156,10 +339,68 @@ export class PageManager {
         this.record(PAGE_VIEW_TYPE, this.createPageViewEvent());
     }
 
+    private moveItemsFromBuffer = (item: any) => {
+        this.page.ongoingActivity.add(item);
+    };
+
     /**
      * Returns true when cookies should be used to store user ID and session ID.
      */
     private useCookies() {
         return navigator.cookieEnabled && this.config.allowCookies;
     }
+
+    /**
+     * Checks whether the virtual page is still being loaded.
+     * If completed:
+     * (1) Clear the timers
+     * (2) Record data using NavigationEvent
+     * (3) Indicate current page has finished loading
+     */
+    private periodicCheckerCallback = () => {
+        if (this.page.ongoingActivity.size === 0 && this.fetchCounter === 0) {
+            clearInterval(this.periodicCheckerId);
+            clearTimeout(this.activityTimeoutCheckerId);
+            this.domMutationObserver.disconnect();
+            this.recordVirtualPageNavigationEvent();
+            this.periodicCheckerId = undefined;
+            this.activityTimeoutCheckerId = undefined;
+            this.page.isLoaded = true;
+        }
+    };
+
+    /** Clears timers and disconnects observer to stop page timing. */
+    private activityTimeoutCallback = () => {
+        clearInterval(this.periodicCheckerId);
+        this.periodicCheckerId = undefined;
+        this.activityTimeoutCheckerId = undefined;
+        this.domMutationObserver.disconnect();
+        this.page.isLoaded = true;
```

Comment:
question: Do we need to reset `fetchCounter` and `ongoingActivity` as well? Do we have a test that does a timeout followed by a completed virtual page load?

Reply:
Synced offline; fetchCounter should not go below zero as long as we do not reset the fetch counter after timeout is declared. This implies that if there's a faulty page load, it will affect subsequent virtual page loads until the fetch/xhr request are completed. If this becomes an issue, we will revise in the future.